# Integrazione finale

Integro tra loro i dati sui titoli presi da wikipedia e quelli presi da IMDb, aggiungo a mano le informazioni su qualche titolo mancante.

In [1]:
import pandas as pd

### Leggo i dataset

Dati integrati con scraping da Wikipedia

In [2]:
# Dataset con titoli e dati di wikipedia

film_wiki = pd.read_csv("Dati_integrati/wiki_film_dic-mar.csv", index_col = 0)
serie_tv_wiki = pd.read_csv("Dati_integrati/wiki_serie_dic-mar.csv", index_col = 0)

Dataset con dati presi da IMDb.

In [3]:
# Dataset con dati imdb

imdb = pd.read_csv("Data_flixpatrol/Dati_netflix_dicembre-aprile/ratings_imdb_dic-mar2021.csv", index_col = 0)

Dati da Flixpatrol: classifiche Netflix giorno per giorno.

In [4]:
# Tutti i dati
netflix = pd.read_csv("Data_flixpatrol/Dati_netflix_dicembre-aprile/netflix_dic-mar2021.csv", index_col = 0)

# Le tre classifiche singole
top10 = pd.read_csv("Data_flixpatrol/Dati_netflix_dicembre-aprile/top_ten_dic-mar2021.csv", index_col= 0)
top_film = pd.read_csv("Data_flixpatrol/Dati_netflix_dicembre-aprile/top_ten_film_dic-mar2021.csv", index_col= 0)
top_serie = pd.read_csv("Data_flixpatrol/Dati_netflix_dicembre-aprile/top_ten_serie_dic-mar2021.csv", index_col= 0)

Creo i due nuovi database __movies__ e __tv_series__ in cui incrocio i dati di Wikipedia con i rating di IMDb.

In [5]:
movies = pd.merge(film_wiki, imdb, left_on = 'Titolo', right_on = 'Title')
movies = movies.drop(columns = ['Title', 'Type'])

In [6]:
tv_series = pd.merge(serie_tv_wiki, imdb, left_on = 'Titolo', right_on = 'Title')
tv_series = tv_series.drop(columns = ['Title', 'Type'])

### Estraggo,correggo i titoli e li ordino in ordine alfabetico

Alcuni titoli presentano delle imperfezioni perché contengono un apostrofo.

In [7]:
titoli_raw = list(set(netflix['name']))  # titoli così come sono nel database netflix

titoli = titoli_raw.copy() # titoli che saranno corretti

In [8]:
len(titoli_raw)

208

In tutto ho 208 titoli. Alcuni titoli nel database originario (contenuti in titoli_raw) sono scritti male, quindi li correggo. titoli_raw conterrà i titoli originari, mentre titoli i titoli corretti.

In [9]:
import re

In [10]:
# Stampo i titoli scritti male (= con un apostrofo)

for t in titoli_raw:
    occ = re.compile('^.*&\#39;.*$')
    
    if occ.match(t):
        print(t)

Finding &#39;Ohana
A Dog&#39;s Way Home
I&#39;m Not a Killer
The Queen&#39;s Gambit
Don&#39;t Listen
Dawson&#39;s Creek
So My Grandma&#39;s a Lesbian!
To All the Boys I&#39;ve Loved Before
Ma Rainey&#39;s Black Bottom


In tutto ho 9 titoli problematici.

In [11]:
# Correggo (in titoli) i titoli scritti male

for t in titoli:
    
    occ = re.compile('(^.*)(\&\#39;)(.*$)')
    found = occ.match(t)
    
    if found:
        titoli.remove(t)
        pulita = re.sub(r'(\&\#39;)',"\'" ,t)
        titoli.append(pulita)

In [12]:
# Ordino in ordine alfabetico

titoli = sorted(titoli)
titoli_raw = sorted(titoli_raw)

In [13]:
# Prova

print(titoli_raw[190])
print(titoli[190])

To All the Boys I&#39;ve Loved Before
To All the Boys I've Loved Before


# Creazione dell'indice di popolarità

Per ogni film f, considerando la classifica totale, il suo indice di popolarità è dato da:

$$ I(f) = \frac{\sum_{g \in G}(10 - ranking(f))}{10*N}$$

dove G è l'insieme dei giorni considerati, N è la cardinalità di G, e il ranking(f) vale 0 se f è primo in classifica, 2 se è secondo eccetera, e infine vale 10 se non è in classifica.
Il ranking vale dunque: $$ranking(f) = ranking - 1$$
dove il ranking è la colonna del dataframe in esame.

In [14]:
lung_titoli = len(titoli_raw)

In [15]:
# In tutto ho 121 giorni (31 (dicembre) + 31 (gennaio) + 28 (febbraio) + 31 (marzo))
num_giorni = 121

In [16]:
# Calcolo indice relativo alla classifica generale (contenuta nel df top10)

def indice_popolarità_gen(riga):
    
    # Stringa con il titolo che considero
    t = riga['Titolo']
    
    for i in range(lung_titoli):
        if titoli[i] == t:
            break

    t_raw = titoli_raw[i]

    # Raggruppo il dataframe delle classifiche per il titolo, così ho i dati relativi ai giorni
    # in cui quel titolo è stato in classifica, ottengo una serie (s) con la posizione (ranking) per ogni 
    # giorno in cui quel titolo è stato in classifica
    s = top10[top10['name'] == t_raw][['name','ranking']]
    
    # Calcolo il numeratore dell'indice
    indice = 0

    for r in s['ranking']:
        indice = indice + (10 - (r - 1)) 
    
    # Divido ora il numeratore per denominatore ottenendo l'indice
    return indice/(10*num_giorni)

Uso la stessa funzione anche per calcolare l'indice con una classifica diversa, cambio solo i df da cui prendo i dati.

In [17]:
# Calcolo indice relativo alla classifica dei film (contenuta nel df top_film)

def indice_popolarità_film(riga):
    
    # Stringa con il titolo che considero
    t = riga['Titolo']
    
    for i in range(lung_titoli):
        if titoli[i] == t:
            break

    t_raw = titoli_raw[i]
    
    # Raggruppo il dataframe delle classifiche per il titolo, così ho i dati relativi ai giorni
    # in cui quel titolo è stato in classifica, ottengo una serie (s) con la posizione (ranking) per ogni 
    # giorno in cui quel titolo è stato in classifica
    s = top_film[top_film['name'] == t_raw][['name','ranking']]
    
    # Calcolo il numeratore dell'indice
    indice = 0

    for r in s['ranking']:
        indice = indice + (10 - (r - 1)) 
    
    # Divido ora il numeratore per denominatore ottenendo l'indice
    return indice/(10*num_giorni)

In [18]:
# Calcolo indice relativo alla classifica delle serie (contenuta nel df top_serie)

def indice_popolarità_serie(riga):
    
    # Stringa con il titolo che considero
    t = riga['Titolo']
    
    for i in range(lung_titoli):
        if titoli[i] == t:
            break

    t_raw = titoli_raw[i]
    
    # Raggruppo il dataframe delle classifiche per il titolo, così ho i dati relativi ai giorni
    # in cui quel titolo è stato in classifica, ottengo una serie (s) con la posizione (ranking) per ogni 
    # giorno in cui quel titolo è stato in classifica
    s = top_serie[top_serie['name'] == t_raw][['name','ranking']]
    
    # Calcolo il numeratore dell'indice
    indice = 0

    for r in s['ranking']:
        indice = indice + (10 - (r - 1)) 
    
    # Divido ora il numeratore per denominatore ottenendo l'indice
    return indice/(10*num_giorni)

Nei due dataframe la colonna _Indice_generale_ si riferisce all'indice i popolarità per quel titolo nella classifica generale, invece _Indice_tipo_ di riferisce all'indice di popolarità per la classifica specifica (solo film o solo serie tv).


Applico ora la funzione al dataframe dei film per creare le nuove colonne 'Indice generale' e 'Indice tipo':

In [19]:
# Dataframe dei film

movies['Indice_generale'] = movies.apply(indice_popolarità_gen, axis = 1)
movies['Indice_tipo'] = movies.apply(indice_popolarità_film, axis = 1)

In [20]:
movies.head(3)

Titolo          Paese  \
0                 365 Days         Poland   
1  5 Is the Perfect Number          Italy   
2   A California Christmas  United States   

                                    Produzione    Distribuzione  \
0  Ekipa Sp. z o.o. Future Space Next Film TVN        Next Film   
1                                   Rai Cinema  01 Distribution   
2                            ESX Entertainment              NaN   

                                                Cast  \
0  Anna-Maria Sieklucka Michele Morrone Bronisław...   
1  Toni ServilloValeria GolinoCarlo BuccirossoIai...   
2                      Lauren Swickard Josh Swickard   

                                            Creatore       Durata  \
0  Tomasz Klimala Barbara Białowąs Tomasz Mandes ...  114 minutes   
1                                              Igort  100 minutes   
2                                    Lauren Swickard  106 minutes   

                          Regista Budget       Box_office  \
0  Barbara Białowąs Tomasz Mandes    NaN  $9.5 million[1]   
1                           Igort    NaN              NaN   
2                 Shaun Piccinino    NaN              NaN   

                                            Sommario  \
0  365 Days (Polish: 365 Dni) is a 2020 Polish er...   
1  5 Is the Perfect Number (Italian: 5 è il numer...   
2  A California Christmas is a 2020 Christmas fil...   

                                               Trama  \
0  Section: Plot (1):\nAfter a meeting between th...   
1  Section: Plot (1):\nIn the 1970s Naples, the r...   
2  Section: Plot (1):\nJoseph Van Aston lives a c...   

                                          Cast_testo  \
0  Section: Cast (1):\nMichele Morrone as Don Mas...   
1           Section: Cast (1):\n\nSubsections (0):\n   
2           Section: Cast (1):\n\nSubsections (0):\n   

                                    Produzione_testo  \
0  Section: Production (1):\nThe film scenes were...   
1                                                NaN   
2                                                NaN   

                                             Critica  Average_rating  \
0  Section: Reception (1):\nThe film made the top...             3.2   
1                                                NaN             5.3   
2                                                NaN             5.7   

   Num_votes  Indice_generale  Indice_tipo  
0    54757.0          0.00000     0.000826  
1     1075.0          0.00000     0.030579  
2     6543.0          0.06281     0.114876

Idem con il dataframe delle serie tv:

In [21]:
# Dataframe delle serie tv

tv_series['Indice_generale'] = tv_series.apply(indice_popolarità_gen, axis = 1)
tv_series['Indice_tipo'] = tv_series.apply(indice_popolarità_serie, axis = 1)

In [22]:
tv_series.head(3)

Titolo          Paese  \
0  Age of Samurai: Battle for Japan            NaN   
1            Agents of S.H.I.E.L.D.  United States   
2               Alice in Borderland          Japan   

                                          Produzione           Distribuzione  \
0                                            Netflix                     NaN   
1  ABC Studios Marvel Television Mutant Enemy Pro...  Walt Disney Television   
2                          Robot Communications Inc.                 Netflix   

                                                Cast  \
0  Masayoshi Haneda Masami Kosaka Hideaki Itō Hay...   
1  Clark Gregg Ming-Na Wen Brett Dalton Chloe Ben...   
2  Kento Yamazaki Tao Tsuchiya Yūki Morinaga Keit...   

                                     Creatore Durata_episodio  \
0                                         NaN   43–45 minutes   
1  Joss Whedon Jed Whedon Maurissa Tancharoen   41–44 minutes   
2                                         NaN   41–52 minutes   

                                             Genere N_stagioni  \
0                                               NaN          1   
1            Action Drama Science fiction Superhero          7   
2  Science fiction[1] Suspense-thriller[2] Drama[3]          1   

                N_episodi                                           Sommario  \
0                       6  Age of Samurai: Battle for Japan is an America...   
1  136 (list of episodes)  Marvel's Agents of S.H.I.E.L.D. is an American...   
2    8 (list of episodes)  Alice in Borderland (Japanese: 今際の国のアリス, Hepbu...   

                                          Cast_testo  \
0                                                NaN   
1  Section: Cast and characters (1):\nClark Gregg...   
2  Section: Cast and characters (1):\n\nSubsectio...   

                                    Produzione_testo  \
0                                                NaN   
1  Section: Production (1):\n\nSubsections (8):\n...   
2  Section: Production (1):\n\nSubsections (3):\n...   

                                             Critica  Average_rating  \
0                                                NaN             7.5   
1  Section: Reception (1):\n\nSubsections (4):\nS...             7.5   
2  Section: Reception (1):\n\nSubsections (2):\nS...             7.7   

   Num_votes  Indice_generale  Indice_tipo  
0     2218.0         0.023967     0.027273  
1   201304.0         0.000000     0.019835  
2    18897.0         0.047934     0.097521

### Esporto in file csv

In [23]:
movies.to_csv("film_integrati.csv")

In [24]:
tv_series.to_csv("serie_tv_integrate.csv")

In [25]:
movies.shape

(140, 19)

In [26]:
tv_series.shape

(68, 18)

## Calcolo i 5 titoli più visti

In [33]:
totale = pd.concat([movies[['Titolo', 'Indice_generale', 'Indice_tipo']].copy(), 
                    tv_series[['Titolo', 'Indice_generale', 'Indice_tipo']]].copy(), ignore_index = True)

In [34]:
totale['Popolarita'] = totale['Indice_generale'] + totale['Indice_tipo']

In [36]:
totale.sort_values(by = 'Popolarita', ascending = False).head(30)

Titolo  Indice_generale  Indice_tipo  \
147                           Bridgerton         0.400826     0.462810   
198                   The Queen's Gambit         0.310744     0.432231   
169                                Lupin         0.281818     0.340496   
161                      Ginny & Georgia         0.259504     0.266942   
174                        New Amsterdam         0.216529     0.247934   
177                            Riverdale         0.207438     0.219008   
159                         Firefly Lane         0.166116     0.183471   
158                  Fate: The Winx Saga         0.147934     0.196694   
180                          Snowpiercer         0.131405     0.208264   
84                           Rose Island         0.125620     0.195041   
144                      Behind Her Eyes         0.137190     0.169421   
200                   Tiny Pretty Things         0.124793     0.161157   
178             SanPa Sins of the Savior         0.127273     0.157025   
196            The Mess You Leave Behind         0.126446     0.153719   
190                            The Crown         0.099174     0.174380   
111   The Christmas Chronicles: Part Two         0.053719     0.176033   
186                              The 100         0.087603     0.139669   
151                            Cobra Kai         0.085124     0.119008   
4                         A Star is Born         0.053719     0.144628   
11                                   Ava         0.067769     0.119008   
203                         Virgin River         0.073554     0.111570   
2                 A California Christmas         0.062810     0.114876   
135                     We Can Be Heroes         0.033058     0.139669   
117                     The Midnight Sky         0.071901     0.091736   
130  To All the Boys: Always and Forever         0.058678     0.104959   
137                              Yes Day         0.051240     0.111570   
73                      Outside the Wire         0.059504     0.103306   
13                        Belleville Cop         0.032231     0.128099   
18                                  Blow         0.031405     0.115702   
155                       Disenchantment         0.052066     0.094215   

     Popolarita  
147    0.863636  
198    0.742975  
169    0.622314  
161    0.526446  
174    0.464463  
177    0.426446  
159    0.349587  
158    0.344628  
180    0.339669  
84     0.320661  
144    0.306612  
200    0.285950  
178    0.284298  
196    0.280165  
190    0.273554  
111    0.229752  
186    0.227273  
151    0.204132  
4      0.198347  
11     0.186777  
203    0.185124  
2      0.177686  
135    0.172727  
117    0.163636  
130    0.163636  
137    0.162810  
73     0.162810  
13     0.160331  
18     0.147107  
155    0.146281

Esporto il dataset finale, con popolarità di film e serie-tv insieme.

In [40]:
totale.to_csv("dataset_popolarita.csv")